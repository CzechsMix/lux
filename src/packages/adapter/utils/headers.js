// @flow
import entries from '../../../utils/entries';

type ChangeType = 'SET' | 'DELETE';
type ChangeData = [string, void | string];
type HandlerChange = (type: ChangeType, data: ChangeData) => void;

class Headers extends Map<string, string> {
  constructor(value: { [key: string]: string }) {
    super(entries(value));
  }

  get(key: string): void | string {
    return super.get(String(key).toLowerCase());
  }

  has(key: string): boolean {
    return super.has(String(key).toLowerCase());
  }

  set(key: string, value: string): this {
    super.set(String(key).toLowerCase(), value);
    return this;
  }

  delete(key: string): boolean {
    return super.delete(String(key).toLowerCase());
  }
}

export class RequestHeaders extends Headers {
  // eslint-disable-next-line no-unused-vars
  set(key: string, value: string): this {
    return this;
  }

  // eslint-disable-next-line no-unused-vars
  delete(key: string): boolean {
    return false;
  }
}

export class ResponseHeaders extends Headers {
  handleChange: HandlerChange;

  constructor(handleChange: HandlerChange) {
    super({});
    this.handleChange = handleChange;
  }

  set(key: string, value: string): this {
    this.handleChange('SET', [key, value]);
    return super.set(key, value);
  }

  delete(key: string): boolean {
    this.handleChange('DELETE', [key]);
    return super.delete(key);
  }
}

Object.defineProperty(ResponseHeaders.prototype, 'handleChange', {
  writable: true,
  enumerable: false,
  configurable: false,
});
