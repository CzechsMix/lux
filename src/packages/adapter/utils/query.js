// @flow
import entries from '../../../utils/entries';
import isObject from '../../../utils/is-object';

const INT = /^\d+$/;
const CSV = /^(?:[\w\d-]+)(?:,[\w\d-]+)*$/;
const NULL = /^null$/;
const BOOL = /^(?:true|false)$/;
const DATE = /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\.\d{3}(Z|\+\d{4})$/;
const TRUE = /^true$/;
const DELIMITER = /[_-\s]+/;

type Params = {
  [key: string]: any;
};

const camelize = (source: string) => (
  source
    .split(DELIMITER)
    .reduce((result, part, idx) => {
      if (part[0]) {
        const [first] = part;

        return (
          result
          + (idx === 0 ? first.toLowerCase() : first.toUpperCase())
          + part.slice(1).toLowerCase()
        );
      }

      return result;
    }, '')
);

const fromString = (source: string): any => {
  if (INT.test(source)) {
    return Number.parseInt(source, 10);
  } else if (BOOL.test(source)) {
    return TRUE.test(source);
  } else if (NULL.test(source)) {
    return null;
  } else if (CSV.test(source)) {
    return source.split(',');
  } else if (DATE.test(source)) {
    return new Date(source);
  }

  return source;
};

export const fromObject = (source: Object): Params => (
  entries(source).reduce((obj, [k, v]) => {
    const result = obj;
    const key = camelize(k);
    let value = v;

    if (typeof value === 'string') {
      value = fromString(value);
    } else if (isObject(value)) {
      value = fromObject(value);
    }

    if (key === 'include' && Array.isArray(value)) {
      value = value.map(camelize);
    } else if (key === 'fields' && isObject(value)) {
      value = entries(value).reduce((fields, [resource, names]) => (
        Object.assign(fields, {
          [resource]: Array.isArray(names) ? names.map(camelize) : names,
        })
      ), {});
    }

    result[key] = value;

    return result;
  }, {})
);
