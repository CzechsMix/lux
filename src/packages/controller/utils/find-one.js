// @flow
import merge from '../../../utils/merge';
// eslint-disable-next-line no-unused-vars
import type { Model, Query } from '../../database';
import type { Request } from '../../request';

import paramsToQuery from './params-to-query';

/**
 * @private
 */
export default function findOne<T: Model>(
  model: Class<T>,
  req: Request
): Query<T> {
  const params = merge(req.defaultParams, req.params);
  const { id, select, include } = paramsToQuery(model, params);

  return model.find(id)
    .select(...select)
    .include(include);
}
