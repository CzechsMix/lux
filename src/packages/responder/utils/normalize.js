// @flow
import { STATUS_CODES } from '../../../constants';
import stringify from '../../../utils/stringify';

import dataFor from './data-for';

type ResponseData = {
  data: string;
  statusCode: number;
};

/**
 * @private
 */
export default function normalize(content: any): ResponseData {
  let data;
  let statusCode;

  switch (typeof content) {
    case 'boolean':
      if (content) {
        statusCode = 204;
      } else {
        statusCode = 401;
        data = dataFor(statusCode);
      }
      break;

    case 'number':
      if (STATUS_CODES.has(content)) {
        statusCode = content;
      } else {
        statusCode = 404;
      }
      data = dataFor(statusCode);
      break;

    case 'object':
      if (content instanceof Error) {
        statusCode = Number.parseInt(content.statusCode, 10) || 500;
        data = dataFor(statusCode, content);
      } else if (!content) {
        statusCode = 404;
        data = dataFor(statusCode);
      } else {
        statusCode = 200;
        data = content;
      }
      break;

    case 'undefined':
      statusCode = 404;
      data = dataFor(statusCode);
      break;

    default:
      statusCode = 200;
      data = content;
  }

  data = stringify(data);

  return {
    data,
    statusCode,
  };
}
