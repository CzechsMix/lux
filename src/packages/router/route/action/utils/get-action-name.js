// @flow
import type { Request } from '../../../../request';

/**
 * @private
 */
export default function getActionName({
  route: {
    action
  }
}: Request): string {
  return action;
}
