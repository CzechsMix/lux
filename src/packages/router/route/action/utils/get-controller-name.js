// @flow
import type { Request } from '../../../../request';

/**
 * @private
 */
export default function getControllerName({
  route: {
    controller: {
      constructor: {
        name
      }
    }
  }
}: Request): string {
  return name;
}
