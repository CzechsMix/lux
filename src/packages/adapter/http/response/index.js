// @flow
import type { ServerResponse } from 'http';

import { ResponseHeaders } from '../../utils/headers';
import type Logger from '../../../logger';
import type { Response } from '../../../response';

export function create(res: ServerResponse, logger: Logger): Response {
  const send = (data: string) => res.end(data);
  const headers = new ResponseHeaders((type, [key, value]) => {
    if (type === 'SET' && value) {
      res.setHeader(key, value);
    } else if (type === 'DELETE') {
      res.removeHeader(key);
    }
  });

  const response = {
    send,
    logger,
    headers,
    end: send,
    stats: [],
    status: (code: number) => {
      // eslint-disable-next-line no-param-reassign
      res.statusCode = code;
      return response;
    },
    getHeader: key => res.getHeader(key),
    setHeader: (key, value) => res.setHeader(key, value),
    removeHeader: key => res.removeHeader(key),
    statusCode: 200,
    statusMessage: 'OK',
  };

  return response;
}
