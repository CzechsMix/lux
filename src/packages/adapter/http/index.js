// @flow
import { createServer } from 'http';
import type { IncomingMessage, ServerResponse } from 'http';

import { PORT, LUX_CONSOLE } from '../../../constants';
import type { AdapterFactory } from '../index';

import { create as createRequest } from './request';
import { create as createResponse } from './response';

const http: AdapterFactory = app => {
  if (!LUX_CONSOLE) {
    const server = createServer(app.exec);

    server.listen(PORT);
  }

  return (req: IncomingMessage, res: ServerResponse) => (
    createRequest(req, app.logger)
      .then(request => [
        request,
        createResponse(res, app.logger),
      ])
      .catch(err => {
        app.logger.error(err);

        // eslint-disable-next-line no-param-reassign
        res.statusCode = 500;
        res.end(err.stack || err.message);
      })
  );
};

export default http;
