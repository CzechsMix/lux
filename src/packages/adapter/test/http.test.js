// @flow
import { IncomingMessage, ServerResponse } from 'http';

import { MIME_TYPE } from '../../jsonapi';
import Logger from '../../logger';
import createAdapter, { request, response } from '../http';
import { getTestApp } from '../../../../test/utils/get-test-app';

const logger = new Logger({
  level: 'ERROR',
  format: 'text',
  filter: {
    params: [],
  },
  enabled: false,
});

describe('module "adapter/http"', () => {
  let app;
  let req;
  let res;
  let adapter;

  beforeAll(async () => {
    app = await getTestApp();
  });

  beforeAll(() => {
    adapter = createAdapter(app);
  });

  describe('#default()', () => {
    beforeEach(() => {
      req = new IncomingMessage({
        encrypted: false,
      });

      Object.assign(req, {
        url: '/posts',
        method: 'GET',
      });

      res = new ServerResponse(req);
    });

    it('resolves with a request/response tuple', async () => {
      expect(await adapter(req, res)).toMatchSnapshot();
    });
  });

  describe('#request', () => {
    describe('#create()', () => {
      let subject;

      describe.skip('- with body', () => {
        beforeEach(async () => {
          const data = JSON.stringify({
            data: {
              type: 'posts',
              attributes: {
                body: '',
                title: 'New Post',
                'is-public': false,
              },
            },
          });

          req = new IncomingMessage({
            encrypted: false,
          });

          Object.assign(req, {
            url: '/posts',
            method: 'POST',
            headers: {
              'Content-Type': MIME_TYPE,
              'Content-Length': data.length,
            },
          });

          const promise = request.create(req, logger);
          const body = Buffer.allocUnsafe(data.length);

          body.write(data);
          req.push(body);

          subject = await promise;
        });

        it('builds a request object from an http.IncomingMessage', () => {
          expect(subject).toMatchSnapshot();
        });
      });

      describe('- without body', () => {
        beforeEach(() => {
          req = new IncomingMessage({
            encrypted: false,
          });

          Object.assign(req, {
            url: '/posts',
            method: 'GET',
          });

          subject = request.create(req, logger);
        });

        it('builds a request object from an http.IncomingMessage', () => {
          expect(subject).toMatchSnapshot();
        });
      });
    });
  });

  describe('#response', () => {
    describe('#create()', () => {
      let subject;

      beforeEach(() => {
        subject = response.create(res, logger);
      });

      it('builds a response object from an http.ServerResponse', () => {
        expect(subject).toMatchSnapshot();
      });
    });
  });
});
