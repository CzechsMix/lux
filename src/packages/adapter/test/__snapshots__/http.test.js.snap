exports[`module "adapter/http" #default() resolves with a request/response tuple 1`] = `
Array [
  Request {
    "defaultParams": Object {},
    "encrypted": false,
    "headers": Array [],
    "logger": Logger {
      "enabled": false,
      "filter": Object {
        "params": Array [],
      },
      "format": "text",
      "level": "WARN",
    },
    "method": "GET",
    "params": Object {},
    "url": Object {
      "auth": null,
      "hash": null,
      "host": null,
      "hostname": null,
      "href": "/posts",
      "params": Array [],
      "path": "/posts",
      "pathname": "/posts",
      "port": null,
      "protocol": null,
      "query": null,
      "search": null,
      "slashes": null,
    },
  },
  Response {
    "end": [Function],
    "getHeader": [Function],
    "headers": Array [],
    "logger": Logger {
      "enabled": false,
      "filter": Object {
        "params": Array [],
      },
      "format": "text",
      "level": "WARN",
    },
    "removeHeader": [Function],
    "send": [Function],
    "setHeader": [Function],
    "stats": Array [],
    "status": [Function],
    "statusCode": 200,
    "statusMessage": "OK",
  },
]
`;

exports[`module "adapter/http" #request #create() - with body builds a request object from an http.IncomingMessage 1`] = `undefined`;

exports[`module "adapter/http" #request #create() - without body builds a request object from an http.IncomingMessage 1`] = `Promise {}`;

exports[`module "adapter/http" #response #create() builds a response object from an http.ServerResponse 1`] = `
Response {
  "end": [Function],
  "getHeader": [Function],
  "headers": Array [],
  "logger": Logger {
    "enabled": false,
    "filter": Object {
      "params": Array [],
    },
    "format": "text",
    "level": "ERROR",
  },
  "removeHeader": [Function],
  "send": [Function],
  "setHeader": [Function],
  "stats": Array [],
  "status": [Function],
  "statusCode": 200,
  "statusMessage": "OK",
}
`;
